File: src/main/java/net/sf/rails/game/model/PriceModel.java
package net.sf.rails.game.model;

import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.financial.Bank;
import net.sf.rails.game.financial.StockSpace;
import net.sf.rails.game.state.GenericState;
import net.sf.rails.game.state.IntegerState;

import java.awt.*;

public final class PriceModel extends ColorModel {

    // FIXME: Remove duplication of company and parent
    private final PublicCompany company;

    /** Share price as represented by a field on the stock market (StockSpace) */
    private final GenericState<StockSpace> stockPrice = new GenericState<>(this, "stockPrice");

    /** Share price as represented by an integer value.
     * Used in 1837 for fixed minor buy cost (face price) in a stock round. */
    private final IntegerState facePrice = IntegerState.create (this, "facePrice");

    private boolean showCoordinates = true;

    private PriceModel(PublicCompany parent, String id, boolean showCoordinates) {
        super(parent, id);
        company = parent;
        this.showCoordinates = showCoordinates;
    }

    public static PriceModel create(PublicCompany parent, String id, boolean showCoordinates) {
        return new PriceModel(parent, id, showCoordinates);
    }

    @Override
    public PublicCompany getParent() {
        return (PublicCompany) super.getParent();
    }

    public void setPrice(StockSpace price) {
        stockPrice.set(price);
    }

    public void setPrice (int price) { facePrice.set(price); }

    public StockSpace getPrice() {
        return stockPrice.value();
    }

    public int getPriceAsInt() { return facePrice.value(); }

    public PublicCompany getCompany() {
        return company;
    }

    @Override
    public Color getBackground() {
        if (stockPrice.value() != null) {
            return stockPrice.value().getColour();
        } else {
            return null;
        }
    }

    @Override
    public Color getForeground() {
        if (getPriceAsInt() == 0) {
            return null;
        } else {
            return Color.GRAY;
        }
    }

    @Override
    public String toText() {
        String text = "";
        if (stockPrice.value() != null) {
            text = Bank.format(getParent(), stockPrice.value().getPrice());
            if (showCoordinates) {
                text += " (" + stockPrice.value().getId() + ")";
            }
        } else if (facePrice.value() > 0) {
            text = Bank.format(getParent(), facePrice.value());
        }
        return text;
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/PriceModel.cs
using GameLib.Net.Game.Financial;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;

namespace GameLib.Net.Game.Model
{
    public class PriceModel : ColorModel
    {
        // FIXME: Remove duplication of company and parent
        private PublicCompany company;

        private GenericState<StockSpace> stockPrice;
        private bool showCoordinates = true;

        private PriceModel(PublicCompany parent, string id, bool showCoordinates) : base(parent, id)
        {
            stockPrice = GenericState<StockSpace>.Create(this, "stockPrice");
            company = parent;
            this.showCoordinates = showCoordinates;
        }

        public static PriceModel Create(PublicCompany parent, string id, bool showCoordinates)
        {
            return new PriceModel(parent, id, showCoordinates);
        }

        new public PublicCompany Parent
        {
            get
            {
                return (PublicCompany)base.Parent;
            }
        }

        public void SetPrice(StockSpace price)
        {
            stockPrice.Set(price);
        }

        public StockSpace GetPrice()
        {
            return stockPrice.Value;
        }

        public PublicCompany Company
        {
            get
            {
                return company;
            }
        }

        override public Color Background
        {
            get
            {
                if (stockPrice.Value != null)
                {
                    return stockPrice.Value.Color;
                }
                else
                {
                    return default(Color);
                }
            }
        }

        override public Color Foreground
        {
            get
            {
                return default(Color);
            }
        }

        
    override public string ToText()
        {
            string text = "";
            if (stockPrice.Value != null)
            {
                text = Bank.Format(Parent, stockPrice.Value.Price);
                if (showCoordinates)
                {
                    text += " (" + stockPrice.Value.Id + ")";
                }
            }
            return text;
        }

    }
}
